src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,416 @@
+{-# LANGUAGE MagicHash       #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RecordWildCards #-}
+{- HLINT ignore "Use camelCase" -}
+
+-- | Incremental, in-memory run consruction
+--
+module Database.LSMTree.Internal.Run.Construction (
+    -- * Types
+    KOp
+  , RawPages (..)
+  , RawKey
+  , RawValue (RawValue, EmptyRawValue)
+  , Operation (..)
+  , BlobRef (..)
+    -- * Page serialisation
+  , serialisePage
+    -- * Incremental page construction
+  , PageIntermediate
+  , pageIntermediateIsEmpty
+  , pageIntermediateEmpty
+  , pageIntermediateAddElem
+  , pageIntermediateSingleton
+    -- * Exposed only for testing
+    -- ** BitMap
+  , BitMap (..)
+  , Bit
+  , emptyBitMap
+  , appendBit
+  , boolToBit
+    -- ** CrumbMap
+  , CrumbMap (..)
+  , Crumb
+  , emptyCrumbMap
+  , appendCrumb
+  , boolsToCrumb
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Bits
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Builder as BB
+import           Data.ByteString.Short.Internal (ShortByteString (SBS))
+import           Data.Foldable
+import           Data.Maybe (isJust)
+import           Data.Primitive.ByteArray
+import           Data.Sequence (Seq (..))
+import qualified Data.Sequence as Seq
+import           Data.Word
+import qualified Database.LSMTree.Internal.Serialise as S
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+-- | A k\/op pair with an optional blob reference.
+type KOp = (RawKey, Operation, Maybe BlobRef)
+
+kopKey :: KOp -> RawKey
+kopKey (x,_,_) = x
+
+kopOp :: KOp -> Operation
+kopOp (_,x,_) = x
+
+kopBlobRef :: KOp -> Maybe BlobRef
+kopBlobRef (_,_,x) = x
+
+--
+-- Pages
+--
+
+
+-- | A string of raw bytes representing one or more disk pages.
+--
+-- Typically, this type only represents a single page, unless it is the result
+-- of adding a larger-than-page value to a run.
+--
+-- Note: the string of bytes is not padded to the target disk-page size.
+data RawPages = RawPages {
+      -- | The number of pages that the output bytestring spans
+      rawPageCount :: !Int
+    , rawPageBytes :: !BS.ByteString
```

Comment:
I should emphasise in a comment that this can't be a lazy bytestring. `RawValue`s point to slices of existing buffers, and we want to copy the bytes from those existing buffers to a new one strictly, or they might be overwritten before they are copied, leading to bogus output